hicommonwealth/commonwealth
# Add $ALEX Eth Token Forum

- Adds $ALEX token as an example ETH Token integration. 
  - Migration layout, and implementation in `App.ts` for chain-loading 
- Adds Token Controller for all token communities
  - Based on contract address in ChainNode model,
  - Assumes ERC20 standard, no extending of the controller for anything other than "V1"
  - If you have the token, you can comment/post in community.
- Adds TokenBalanceCache on server to check balances of posters in token forums.
 
#### TODOs
- ~~Server-side validation/cache of last time we checked you have the required token to participate.~~
- Threshold of participation? Currently set to `!balance.isZero()`
- ~~Unit testing for TokenBalanceCache / modified server calls.~~

## Motivation and Context
Allow us to more rapidly add communities by contract, by creating a basic controller and template migration.

## How has this been tested?
Run `yarn reset-server` and that will create an "Alex (ropsten)" chain, in which the account with private key 1111..1111 (public key: 0x19E7E376E7C213B7E7e7e46cc70A5dD086DAff2A) has tokens on Ropsten.
- Add this account to metamask and the flow should work as expected (i.e. you should be able to post, commend, react).
- Switch to a different account and these features should be disabled on the UI.
- Remove the UI disables manually and the server should reject the requests.

## Have proper tags been added (for bug, enhancement, breaking change)?

## Does this PR affect any server routes?
- [x] yes, but I did not run tests
